### PR TITLE
[WIP] Refactor

### PR DESCRIPTION
**Pull request by tangrufus**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -1,36 +1,54 @@
...
+            "dev-master": "2.x-dev"
+        }
+    },
+    "require-dev": {
+        "codeception/codeception": "^4.1",
+        "codeception/module-asserts": "^1.0.0",
+        "codeception/mockery-module": "^0.4.0"
+    },
+    "scripts": {
+      "test": "codecept run"
```
Comment (austinpray) [inline composer.json:52]: We already have a WIP unit test branch here https://github.com/roots/bedrock-autoloader/pull/17

**Comment by vinkla** [comment]
What are the benefits to this approach instead of the current state? Does this make the autoloader more reliable?

**Comment by austinpray** [comment]
> What are the benefits to this approach instead of the current state? Does this make the autoloader more reliable?

Would also like a rundown of this as well. From the code I can guess a lot of this is aimed at making things more unit testable? Honestly besides validating some extremely basic logic like in https://github.com/roots/bedrock-autoloader/pull/17 I'm mostly concerned with e2e + integration testing here

**Comment by tangrufus** [comment]
One of my goal is to support mu plugins out of [`'/../' . basename(WPMU_PLUGIN_DIR)`](https://github.com/roots/bedrock-autoloader/blob/9ef75905ff9ff367a2b277a9ee58a6b3c76c9a9c/src/Autoloader.php#L45).

With https://github.com/TangRufus/bedrock-autoloader/blob/80cf4c061abcd9bbbaddfb67ab3d5aee8e009539/src/Autoloader.php#L32 we open up the possibility to use [custom plugin repos](https://github.com/TangRufus/bedrock-autoloader/blob/80cf4c061abcd9bbbaddfb67ab3d5aee8e009539/src/PluginRepo/PluginRepoInterface.php).

Use case: Load mu plugins outside web root (i.e: `bedrock/web`). 
MU plugins can now be installed at all of these places:
- bedrock/vendor/my-vendor
- bedrock/web/app/mu-plugins
- [somewhere/unrelated](https://github.com/roots/bedrock-autoloader/issues/12)

---

Re: Unit tests

I will change to use #17 if we continue with this pr

---

What are your opinions?

**Comment by vinkla** [comment]
> What are your opinions?

I think it is great that this pull request makes it possible to move the autoloader out of the `mu-plugin` directory. I also think you're doing awesome work pushing this project forward!

My concern is that the new implementation could become harder to understand. If one compares the current implementation with the new:

- 10 classes compared to 1 (current implementation).
- 3 new packages installed with Composer.

When I looked through the code base, I thought it was hard to undertand how the bits and pieces were connected. If I were a new developer looking at the project and wanted to know how it works, I might give up.

Is there a way to achieve the same result while maintaining the simplicity of the current implementation?

**Comment by tangrufus** [comment]
Over-engineering. Forget me.

**Pull request closed without merging**